src/treetensornetworks/projttns/projttn_apply.jl

```diff
@@ -1,57 +1,114 @@
-  return ProjTTNApply(P.pos, copy(P.psi0), copy(P.H), copy(P.environments))
+  return ProjTTNApply(
+    pos(P), copy(init_state(P)), copy(operator(P)), copy_keys_values(environments(P))
+  )
 end
 
+function unsafe_copy(P::ProjTTNApply)
+  return ProjTTNApply(P.pos, copy(init_state(P)), copy(operator(P)), copy(environments(P)))
+end
 function set_nsite(P::ProjTTNApply, nsite)
   return P
 end
 
 function shift_position(P::ProjTTNApply, pos)
-  return ProjTTNApply(pos, P.psi0, P.H, P.environments)
+  return ProjTTNApply(pos, init_state(P), operator(P), environments(P))
+end
+
+function set_environments(p::ProjTTNApply, environments)
+  return ProjTTNApply(pos(p), init_state(p), operator(p), environments)
+end
+
+function set_environment(p::ProjTTNApply, edge, env)
+  newenv = merge(p.environments, Dictionary((edge,), (env,)))
+  return ProjTTNApply(pos(p), init_state(p), operator(p), newenv)
 end
 
+set_environment!(p::ProjTTNApply, edge, env) = set!(environments(P), edge, env)
+
 function make_environment!(
-  P::ProjTTNApply{V}, psi::TTN{V}, e::NamedEdge{V}
+  P::ProjTTNApply{V}, state::AbstractTTN{V}, e::NamedEdge{V}
 )::ITensor where {V}
   # invalidate environment for opposite edge direction if necessary
-  reverse(e) ∈ incident_edges(P) || unset!(P.environments, reverse(e))
+  reverse(e) ∈ incident_edges(P) || unset!(environments(P), reverse(e))
   # do nothing if valid environment already present
-  if haskey(P.environments, e)
+  if haskey(environments(P), e)
     env = environment(P, e)
   else
     if is_leaf(underlying_graph(P), src(e))
       # leaves are easy
-      env = P.psi0[src(e)] * P.H[src(e)] * dag(psi[src(e)])
+      env = init_state(P)[src(e)] * operator(P)[src(e)] * dag(state[src(e)])
     else
       # construct by contracting neighbors
       neighbor_envs = ITensor[]
       for n in setdiff(neighbors(underlying_graph(P), src(e)), [dst(e)])
-        push!(neighbor_envs, make_environment!(P, psi, edgetype(P)(n, src(e))))
+        push!(neighbor_envs, make_environment!(P, state, edgetype(P)(n, src(e))))
       end
       # manually heuristic for contraction order: two environments, site tensors, then
       # other environments
       frst, scnd, rst = _separate_first_two(neighbor_envs)
-      itensor_map = vcat(P.psi0[src(e)], frst, scnd, P.H[src(e)], dag(psi[src(e)]), rst)
+      itensor_map = vcat(
+        init_state(P)[src(e)], frst, scnd, operator(P)[src(e)], dag(state[src(e)]), rst
+      ) # no prime here in comparison to the same routine for Projttn
       # TODO: actually use optimal contraction sequence here
       env = reduce(*, itensor_map)
     end
     # cache
-    set!(P.environments, e, env)
+    set_environment!(P, e, env)
   end
   @assert(
-    hascommoninds(environment(P, e), psi[src(e)]),
+    hascommoninds(environment(P, e), state[src(e)]),
     "Something went wrong, probably re-orthogonalized this edge in the same direction twice!"
   )
   return env
 end
+
+function make_environment(
```

Comment:
Again, I would prefer to not have both `make_environment` and `make_environment!`.

Reply:
Ok. Do you have an opinion of whether the mutating or non mutating version is prefered? With the copying behaviour fixed it's reasonable to have position do a copy of its input projected_operator and call mutating functions from there. But we can also do full non-mutating calls to the end.

Reply:
Didn't see your comment above that answers this.

Reply:
It's a good question, and I don't think it's obvious one way or another without benchmarking. But let's lean towards non-mutating versions for now and add back mutation as needed for performance.